new: usr: Support file-based scheduler configuration.

ashmastaflash opened the pull request:
Introduced file-based configuration for the scheduler.
Configuration files are in 'ini' format. An
example configuration file is named 'example.conf',
and is located in the root of this repository.

Removed the instance variables which are related to
environment variables from Containerized()
class. These variables will now be set when
each container is run, based on the env_expand
argument which has been introduced into the generic
method for launching containers.

Refactored firewall graph generator and ec2 footprint
reporter functions into donbot project, which now will
use the generic container launch functionality to
perform these functions.

Celery version updated to 4.2.0rc4, to avoid issues with
included pyamqp library in celery version 4.1.

Closes #21

Closes #23

Closes #26

eatwithforks commented on apputils/config_validator.py:
```diff
@@ -0,0 +1,52 @@
+    def validate_config(cls, config):
+        """Validate required fields are present and module is 'scheduler'."""
+        msg = ""
+        if ("service" not in config
```
pull this if statement into a function so it can be separately tested when there are missing services, etc. ie. if self.missing_services? return

eatwithforks commented on apputils/config_validator.py:
```diff
@@ -0,0 +1,52 @@
+                         'day_of_month', 'month_of_year']
+
+    @classmethod
+    def validate_config(cls, config):
```
pull this if statement into a function so it can be separately tested when there are missing services, etc. ie. if self.missing_services? return

separate the rest of missing sections logic into functions that can be individually tested

mong2 commented on apputils/config_validator.py:
```diff
@@ -0,0 +1,52 @@
+        if missing_schedule_fields:
+            msg += ("ConfigValidator: Mising from [schedule] section: %s\n"
+                    % ", ".join(missing_schedule_fields))
+        if config["task_config"]["read_only"] not in ["true", "false"]:
```
should utilize `getboolean`

mong2 commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+            config_path(str): Path to configuration directory.
+
+        """
+        files_in_dir = [os.path.join(config_path, f)
```
this should be a separate function

ashmastaflash replied:
done.

mong2 commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+        files_in_dir = [os.path.join(config_path, f)
+                        for f in os.listdir(config_path)
+                        if os.path.isfile(os.path.join(config_path, f))]
+        for target in sorted(files_in_dir):
```
why are we sorting `files_in_dir`

ashmastaflash replied:
We sort the files so that they're loaded in a predictable order.  Makes behavior more deterministic.

mong2 replied:
why not use `ordereddict`?

ashmastaflash replied:
Because it's not a dictionary, it's a list.  There's no need for key-value pairs in this part of the code. Using `ordereddict` would still require sorting the contents, if you want the contents in alpha order.
https://docs.python.org/2/library/collections.html#collections.OrderedDict

mong2 commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+                task_name = config["task_config"]["task_name"]
+                self.scheduled_tasks[task_name] = config
+                # Coerce `read_only` to boolean
+                if self.scheduled_tasks[task_name]["task_config"]["read_only"] == "true":  # NOQA
```
should utilize `getboolean`

ashmastaflash replied:
done

eatwithforks commented on apputils/config_validator.py:
```diff
@@ -0,0 +1,52 @@
+            return False
+        missing_sections = [x for x in cls.sections_required
+                            if x not in config]
+        if missing_sections:
```
Can you utilitize https://docs.python.org/2/library/configparser.html instead of using manual check

exception ConfigParser.NoOptionError¶
Exception raised when a specified option is not found in the specified section.

mong2 commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+        config = ConfigParser.RawConfigParser({}, dict)
+        config.optionxform = str
+        with open(config_file_path, 'r') as conf_file:
+            config.readfp(conf_file)
```
> readfp(fp, filename=None)
> Deprecated since version 3.2: Use read_file() instead.
>

ashmastaflash replied:
@mong2 that function is not available in Python2.7.

eatwithforks commented on apputils/config_validator.py:
```diff
@@ -0,0 +1,52 @@
+"""Validator for config files."""
+
+
+class ConfigValidator(object):
```
Should have tests for all functions in here. For configparser stuff, use its native exceptions for hte assertions

ashmastaflash replied:
Uses configparser native functionality

mong2 commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+                    self.scheduled_tasks[task_name]["task_config"]["read_only"] = True  # NOQA
+                else:
+                    self.scheduled_tasks[task_name]["task_config"]["read_only"] = False  # NOQA
+        self.print_tasks(self.scheduled_tasks)
```
this function should just return `self.scheduled_tasks`. Another function can call `self.print_tasks`. With this way, it would be easier to add a `debug` flag in the future

ashmastaflash replied:
Done.

eatwithforks commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+        """Return dictionary that describes celerybeat tasks, from config."""
+        beats = {}
+        for j in self.scheduled_tasks.items():
+            beats[j[0]] = {'task': 'halocelery.tasks.generic_bound_containerized_task',  # NOQA
```
move this logic to a function. pass in J to a builder function so you can test building the hash correctly from a single scheduled task item. can also assert each item/value is expected

ashmastaflash replied:
Done

eatwithforks commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+        return beats
+
+    @classmethod
+    def format_task(cls, task):
```
Can all functions that specifically format and prints stuff be moved to a separate class? It's confusing to have formatting functions scattered through each file. should be consolidated.

ashmastaflash replied:
Done.

eatwithforks commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+    def beat_tasks_from_config(self):
+        """Return dictionary that describes celerybeat tasks, from config."""
+        beats = {}
+        for j in self.scheduled_tasks.items():
```
what is j[0] and j[1]? can you make the naming more clear and is it possible to use k,v instead of positionings.

ashmastaflash replied:
Done, refactored

mong2 commented on apputils/config_manager.py:
```diff
@@ -0,0 +1,83 @@
+    def beat_tasks_from_config(self):
+        """Return dictionary that describes celerybeat tasks, from config."""
+        beats = {}
+        for j in self.scheduled_tasks.items():
```
please use key, value pair to loop through the dictionary. It is not clear what j[0] and j[1] stands for.

ashmastaflash replied:
Done

mong2 commented on tasks.py:
```diff
@@ -86,35 +65,97 @@ def remove_ip_from_list(ip_address, ip_zone_name):
+                                                    env_expand.copy(),
+                                                    read_only)
+        print("TaskRunner: %s" % finished_msg)
     except Exception as e:
```
could we please specify the error we are catching here? it's not efficient to retry on all errors

mong2 commented on tasks.py:
```diff
@@ -86,35 +65,97 @@ def remove_ip_from_list(ip_address, ip_zone_name):
-        'schedule': crontab(hour=scans_hour, minute=scans_min),
-        'args': (os.getenv("SCANS_S3_BUCKET"), )}
-    }
+        print("TaskRunner: %s" % fail_msg)
```
i would suggest to utilize the `import sys` to do logging. It provides the ability to do std.out or std.err

ashmastaflash replied:
logging consolidated.  See newest commit notes.

ashmastaflash committed:
new: usr: Support file-based scheduler configuration.

Introduced file-based configuration for the scheduler.
Configuration files are in 'ini' format. An
example configuration file is named 'example.conf',
and is located in the root of this repository.

Removed the instance variables which are related to
environment variables from Containerized()
class. These variables will now be set when
each container is run, based on the env_expand
argument which has been introduced into the generic
method for launching containers.

Refactored firewall graph generator and ec2 footprint
reporter functions into donbot project, which now will
use the generic container launch functionality to
perform these functions.

Introduced Utility.log_stdout and Utility.log_stderr to
consolidate log message printing.

Celery version updated to 4.2.0rc4, to avoid issues with
included pyamqp library in celery version 4.1.

Closes #21

Closes #23

Closes #26
